src/Processors/QueryPlan/PartsSplitter.h

```diff
@@ -0,0 +1,322 @@
+                //                                                                                                          ^
+                //                                                                                                          +---------------------- split_mark
+                //
+                // current_layer (Filter: borders[i-1] <= pk < borders[i]):                  |---|---|---               |---|---
+                //
+                //    next_layer (Filter: borders[i] <= pk < borders[i+1]):                                                 |---|---|--...--|---
+                //
+                // Every already started but not yet finished range should be divided into two subranges:
+                // one will go to the current layer and should contain all rows with primary key values less than border
+                // and another one with all rows with PK values not less than border should be returned back into the queue.
+                for (const auto & range : started_ranges)
+                {
+                    const auto part_idx = range.part_idx;
+                    const auto split_mark = splitByBorder(part_idx, range, event.value);
+                    result_layers.back().emplace_back(
+                        parts[part_idx].data_part, part_idx, MarkRanges{{range.begin, std::min<size_t>(split_mark + 1, range.end)}});
+                    const auto subrange_for_next_layer = MarkRange{split_mark, range.end};
+                    if (subrange_for_next_layer.begin < subrange_for_next_layer.end)
+                    {
+                        auto new_begin_value = index_access->getValue(part_idx, subrange_for_next_layer.begin);
+                        event_queue.emplace(std::move(new_begin_value), part_idx, subrange_for_next_layer);
+                    }
+                }
+                started_ranges.clear();
+                result_layers.emplace_back();
+            }
+        }
+        for (const auto & range : started_ranges)
+            result_layers.back().emplace_back(parts[range.part_idx].data_part, range.part_idx, MarkRanges{range});
+
+        for (size_t layer = 0; layer < result_layers.size(); ++layer)
+        {
+            LOG_DEBUG(&Poco::Logger::get("PartsSplitter"), "Layer {}, ranges:", layer);
+            for (const auto & range : result_layers[layer])
+                for (const auto & mark_range : range.ranges)
+                    LOG_DEBUG(
+                        &Poco::Logger::get("PartsSplitter"),
+                        "part_idx={}, begin={}, end={}",
+                        range.part_index_in_query,
+                        mark_range.begin,
+                        mark_range.end);
+        }
+
+        return result_layers;
+    }
+
+    /// Will return borders.size()+1 filters in total, i-th filter will accept values from the range [borders[i-1], borders[i]).
+    std::vector<ASTPtr> buildFilters(const std::vector<IndexValue> & borders)
```

Comment:
I think ASTPtr here can be a little bit redundant.
Can't we add special filter steps which can properly process IndexValue?